### ISSUE TITLE
Problem with DerivativeForm Constructor

### ISSUE THREAD

**Issue by rgoermer** [dealii/dealii#14748]
I am trying to understand the proper instantiation of a `DerivativeForm<2, dim, spacedim, Number>`; as far as I understood it can be interpreted as a rank 3 tensor describing the hessian **HF** of a mapping **F**. And from the documentation it is mentioned that it can be thought of a matrix with dimensions `(spacedim x dim x dim)`. 

Then, when using `DerivativeForm<2, dim spacedim, Number>::operator[](const unsigned int)` an object of type `Tensor<2, dim, Number>` is returned. In other words, we get a `dim x dim` matrix written as a rank 2 tensor. 

However, for the constructor we hand an object of type (suppose `dim != spacedim`) `Tensor<2, spacedim, Tensor<1,  dim> >` to build **HF** out of. Using `Tensor<2, spacedim, Tensor<1, dim>>::operator[](const unsigned int)`, an object of type `Tensor<1, spacedim, Tensor<1, dim>>` will be returned for that object, that represents a rank 2 tensor with dimensions `spacedim x dim`. 

In other words, the object given to the constructor of `DerivativeForm<2, dim, spacedim, Number>` of type `Tensor<2, spacedim, Tensor<1,  dim> >` seems to describe a rank 3 tensor similar to a `(spacedim x spacedim x dim)`-matrix. This is also correct from the documentation:

>  A rank-2 tensor is a linear operator that can transform a vector into another vector. These tensors are similar to matrices with dim×dim components.

This leads me to believe, that there is a dimension mismatch for `DerivativeForm<order, dim, spacedim>`, whenever `dim != spacedim`, and I suppose the correct object to hand to the constructor would be `Tensor<1, spacedim, Tensor<order,  dim> >` at least for `order = 2`. 

With the above, I still try to comprehend why the compiler returns no errors at this point: 
```
template <int order, int dim, int spacedim, typename Number>
 inline DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
   const Tensor<order, spacedim, Tensor<1, dim, Number>> &T)
 {
   for (unsigned int j = 0; j < spacedim; ++j)
     (*this)[j] = T[j];
 }
```
In my opinion the data types on the left- and right-hand-side are incompatible, but it compiles just fine. 

Is there anyone, who might shed some light on this topic? 

Edit: The more I think about it, the desired input just makes no sense: Mathematically, we can think of the Hessian of a function F : \R^dim \to \R^spacedim as a rank 3 tensor **HF** with dimensions `spacedim x dim x dim` as mentioned in the documentation, where we collect for each F_i the hessian **HF_i**. The data type seems to relate to a `Tensor<1, spacedim, Tensor<2, dim> >`. On the other hand, we can also identify it with a `dim x dim` matrix, where each entry is a `spacedim` vector, i.e. **HF** can be viewed as a `dim x dim x spacedim` matrix, where we can use the relating data type `Tensor<2, dim, Tensor<1, spacedim> >`. In this case the sizes `dim` and `spacedim` are switched compared to the needed type `Tensor<2, spacedim, Tensor<1, dim> >`.

**Comment by drwells** [comment]
Interesting question - we should document these restrictions better - some of these functions, as far as I can tell, don't make sense with nonzero codimension. In particular, if I recall correctly, this constructor was added to support some other matrix-free code (which does not support nonzero codimension problems yet).

If I try to compile
```cpp
      constexpr int order    = 2;
      constexpr int dim      = 2;
      constexpr int spacedim = 3;

      Tensor<order, spacedim, Tensor<1, dim>> t;

      auto d2 = DerivativeForm<order, dim, spacedim>(t);
```

I get the correct compilation error
```
/home/drwells/Documents/Code/CPP/dealii-dev/include/deal.II/base/derivative_form.h: In instantiation of 
    'dealii::DerivativeForm<order, dim, spacedim, Number>::DerivativeForm(
        const dealii::Tensor<order, spacedim, dealii::Tensor<1, dim, Number> >&) 
    [with int order = 2; int dim = 2; int spacedim = 3; Number = double]':
/home/drwells/Documents/Code/CPP/dealii-experiments/derivative-form-bug/bug.cc:34:55:   required from here
/home/drwells/Documents/Code/CPP/dealii-dev/include/deal.II/base/derivative_form.h:205:16: error: 
    no match for 'operator=' (operand types are 'dealii::Tensor<2, 2, double>' and 
    'const dealii::Tensor<2, 3, dealii::Tensor<1, 2, double> >::value_type' 
    {aka 'const dealii::Tensor<1, 3, dealii::Tensor<1, 2, double> >'})
  205 |     (*this)[j] = T[j];
      |     ~~~~~~~~~~~^~~~
```

Since it's a template class, compilation won't fail until one tries to call that constructor with the 'wrong' types. Were you able to get farther than this? I believe this should fail to compile in all nonzero codimension cases.

**Comment by rgoermer** [comment]
I hadn't actually tried to compile a test program like you did; I just assumed it would work since it's included in the library. The error you get is exactly what I expected to get when compiling deal.II. But I guess since this function is nowhere to be called when building deal.II (with e.g. candi) it is just fine. 

Can you elaborate the codimension aspect? I'm not quite familiar with that concept. I assume the codimension in this case is 1, as 3 - 2 = 1, when viewing the parametric space (dim) as sub-space from the physical space (spacedim)?

**Comment by rgoermer** [comment]
Ok, I see now. I still don't understand why is it important for this class. 

Moreover, setting `spacedim = 2` in your code compiles just fine [as you suggested]. I would have expected a similar error in this case [with zero codimension], as the right-hand side corresponds to a `Tensor<1, 2, Tensor<1, 2> >` and the left-hand side corresponds to `Tensor<2, dim>`. Surely, they are similar to each other, but still different data types. Is there a special way the assignment operator is written, s.t. it works?

**Comment by rgoermer** [comment]
I have now moved on with this problem by simply not defining my hessians as `DerivativeForm<2, dim, spacedim>` but instead using `Tensor<1, spacedim, Tensor<2, dim>>` as it essentially stores the type I want and that is described in the documentation. That is fine for me, as I probably will not need the computation of norms, covariants and others for the hessian. 

Maybe the Developers of deal.ii will allow this case at some point in the future. In my opinion it is sufficient to simply change the input from `Tensor<order, spacedim, Tensor<1,dim>>` to `Tensor<1, spacedim, Tensor<order, dim>>` to allow also non-zero codimensions.

**Comment by bangerth** [comment]
At the very least to explain what *should* happen:

* This class represents the derivatives of functions `f : R^spacedim -> R` where their domain is `dim`-dimensional. These functions live on surfaces of dimension `dim` embedded in `spacedim` dimensional space, say, on the surface of the earth (dim=2, spacedim=3). Of course, if `dim=spacedim`, then they just live in the ambient space, say a function that is defined for the entire Earth interior.
* Derivatives are taken with regard to the `dim` variables that parameterize the domain -- i.e., a set of unit vectors of the tangent plane.
* So the first derivative is a `dim x spacedim` or `spacedim x dim` matrix, where we just need to choose a convention.
* Similarly, the second derivative is going to be a `dim x dim x spacedim` or `spacedim x dim x dim` matrix, depending on convention.

For second derivatives, it therefore makes sense to represent the `spacedim x dim x dim` matrix as a vector (with `spacedim` elements) of `dim x dim` symmetric matrices, and the creators of this class have chosen to express this as a `Tensor<1,spacedim, T>` outer vector with `T=Tensor<2,dim>` inner symmetric matrix. It is perhaps not crazy to suggest that `std::array<Tensor<2,dim>,spacedim>` is a better choice (patch gladly accepted!) but the `std::array` class wasn't available when this class was originally written.

In other words, I think that the type in the constructor is correct, though the documentation may not be and we will gladly accept suggestions how to improve it!

**Comment by bangerth** [comment]
I read through this again and was wrong on most counts :-) I edited the first bullet point above to make it right. As for the constructor signature: It is currently `Tensor<order, spacedim, Tensor<1, dim, Number>>`, which indeed looks incorrect. It should have been `Tensor<1, spacedim, Tensor<order, dim, Number>>` instead, or better `std::array<Tensor<order, dim, Number>,spacedim>`. I will see whether that is something we can fix.
